GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,240 @@
+        {
+            if (healthData.GitTrackedFilesCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)healthData.PlaceholderCount / healthData.GitTrackedFilesCount;
+        }
+
+        public decimal GetModifiedPathsPercentage(GVFSEnlistmentHealthData healthData)
+        {
+            if (healthData.GitTrackedFilesCount == 0)
+            {
+                return 0;
+            }
+
+            return (decimal)healthData.ModifiedPathsCount / healthData.GitTrackedFilesCount;
+        }
+
+        /// <summary>
+        /// Take a string representing a file path on system and pull out the upper-most directory from it as a string, or GVFSConstants.GitPathSeparator if it is in the root
+        /// </summary>
+        /// <param name="path">The path to a file to parse for the top level directory containing it</param>
+        /// <returns>A string containing the top level directory from the provided path, or GVFSConstants.GitPathSeparator if the path is for an item in the root</returns>
+        private string ParseTopDirectory(string path)
+        {
+            int whackLocation = path.IndexOf(GVFSConstants.GitPathSeparator);
+            if (whackLocation == -1)
+            {
+                return GVFSConstants.GitPathSeparatorString;
+            }
+
+            return path.Substring(0, whackLocation);
+        }
+
+        /// <summary>
+        /// Categorizes a list of paths given as strings by mapping them to the top level directory in their path
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a GVFSConstants.GitPathSeparator, then that means it falls within the root
+        /// However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it counts towards itself for hydration
+        /// </remarks>
+        /// <param name="paths">An enumerable containing paths as strings</param>
+        /// <param name="directoryTracking">A dictionary used to track the number of files per top level directory</param>
+        /// <param name="parentDirectory">Paths will only be categorized if they are descendants of the parentDirectory</param>
+        /// <param name="isFile">Flag which specifies whether the list of paths are paths to files (true) or directories (false)</param>
+        private int CategorizePaths(IEnumerable<string> paths, Dictionary<string, int> directoryTracking, string parentDirectory, bool isFile)
+        {
+            int count = 0;
+            foreach (string path in paths)
+            {
+                // Ensure all paths have the same slashes
+                string formattedPath = path.Replace('\\', GVFSConstants.GitPathSeparator);

```

Comment:
Should we normalize paths up front? so this method can assume all paths are already in the normalize format?

Reply:
My thoughts were that after the class is constructed, as soon as anything is ready to touch any data, specifically the `CalculateStatistics()` method, the first thing that happens to all of the paths is they all pass through this function (`CategorizePaths()`). Except for perhaps some crazy edge case, paths are already standardized. I figure that given that this method is responsible for categorizing the paths, normalizing them also seems like a reasonable operation. I'd love to hear your thoughts on this!

Reply:
The other option would be that if we switch over to a class responsible for all of the path data, there could be a `Normalize()` method in it which would iterate over all of the paths and normalize them beforehand